### PR TITLE
(Minimum) Changes to avoid floating-point exceptions

### PR DESCRIPTION
**Pull request by pramodk**
* If coreneuron is compiled with MPI but run without --mpi
        flag then coreneuron crashes. This doesn't allow NEURON
        to run existing tests (e.g. using python API)
* Instead of using MPI function, using wrappers from mpispike.cpp
* Updated all reduce wrapper to accept size_t / unsigned long int than just size_t
* Update cast and summary printf

Further cleanup will be done in separate PR, see #490

**How to test this?**

On BB5

```bash
module load unstable hpe-mpi cmake python-dev flex bison cmake gcc
cmake ..
make -j8
make test
```
Specifically look at:

```
ctest -R ring_serial -VV
```

**Test System**
 - Specifically HPE-MPI aka MPT

CI_BRANCHES:NEURON_BRANCH=master,

### CODE REVIEW

**Review by ohm314 — APPROVED** [review]
LGTM

File: coreneuron/io/mech_report.cpp [hunk 1]
```diff
@@ -19,7 +19,7 @@ namespace coreneuron {
 void write_mech_report() {
     /// mechanim count across all gids, local to rank
     const auto n_memb_func = corenrn.get_memb_funcs().size();
-    std::vector<unsigned long long> local_mech_count(n_memb_func, 0);
+    std::vector<size_t> local_mech_count(n_memb_func, 0);
```
Comment (olupton) [inline coreneuron/io/mech_report.cpp:22]: It seems a bit messy to use `size_t` here and then make `nrnmpi_long_allreduce_vec` expect `unsigned long` when `nrnmpi_long_allreduce` expects `long`.

Might it be easier/more consistent to just use `long` and sidestep issues like whether `size_t*` -> `unsigned long*` is OK, and whether `MPI_UNSIGNED_LONG_LONG` is the correct value for `size_t`?

Comment (pramodk) [reply coreneuron/io/mech_report.cpp:22]: I did somewhat sloppy job here. I started changing stuff and quickly I realised that types are used inconsistently at multiple places. So I end-up creating #491.

> Might it be easier/more consistent to just use long and sidestep issues like whether

Considering these routines are mostly used for statistics printing and `long` should be sufficiently large (on 64bit OS), I will replace size_t with long as you suggested.

**Review by olupton — APPROVED** [review]
LGTM!

**Comment by pramodk** [comment]
retest this please

**Pull request merged**